src/sage/categories/finite_dimensional_algebras_with_basis.py

```diff
@@ -181,13 +184,14 @@ def product_on_basis(i, j):
                         return B[i] * B[j]
 
                 keys = B.keys()
-                cache = [{(i, j): c for i in keys for j, c in product_on_basis(y, i)}
+                cache = [{(i, j): c for i in keys
+                         for j, c in product_on_basis(y, i).monomial_coefficients().items()}
                          for y in keys]
-                mat = [[sum(x.get((j, i), 0) * c for (i,j), c in y.items())
+                mat = [[F.sum(x.get((j, i), 0) * c for (i,j), c in y.items())
```

Comment:
```suggestion
                mat = [[F.sum(x.get((j, i), F.zero()) * c for (i,j), c in y.items())
```
Although, it probably would be better just to skip that `c` altogether...

Reply:
Good suggestion, but let me use `zero = F.zero()` to avoid rebuilding it again and again.